src/ni/vsbuild/Pipeline.groovy

```diff
@@ -196,6 +198,30 @@ class Pipeline implements Serializable {
             if (script.fileExists(archiveLocation)) {
                script.failBuild("Refusing to build, $archiveLocation already exists and would be overwritten.")
             }
+
+            // If build was not caused by an upstream job, i.e. it was started manually
+            // or because of a source change, always build it.
+            // Note that the Jenkins CI lists this method as EXPERIMENTAL, but it has worked for many years.
+            def upstreamBuild = script.currentBuild.getBuildCauses('hudson.model.Cause$UpstreamCause')
+            if (!upstreamBuild) {
+               return true
+            }
+
+            // If build was started by an upstream job, check if there have been any
+            // changes to the repo since the last successful build.
+            // If there are changes, do the build.
+            // If not, tell the upstream job to use the previous build artifacts.
+            def lastBuildLocation = script.findLatestDirectory(archiveParentLocation)
+            def rebuild = script.needsRebuild(lastBuildLocation, commit, pipelineInformation.lvVersions)
+            if (rebuild.toLowerCase() == 'false') {
```

Comment:
`needsRebuild.groovy` should probably handle the conversion from string to boolean, and return a native Java bool.

Checking for a string here seems like a violation of the encapsulation.